Support for downsampling and kicking out short scenes

sssangha opened the pull request:
To allow for downsampling and kick out of short scenes, the following changes were made:
1. Use the "-cr" flag in ariaTSsetup.py and ariaExtract.py to specify a percentage to resample the final outputs wrt the original resolution of a standard product. DEM/mask products are also resampled accordingly, but only after all other layers have already been extracted in order to avoid introducing artifacts when downsampling masked arrays or extracting metadata layers with a downsampled DEM. DEM and mask objects are now tagged with their filenames to reduce complexity when trying to access these names during the resampling stage.
2. The new "resampleRaster" function in vrtManager.py handles the resampling calls.
3. Use the "-po" flag in ariaTSsetup.py and ariaExtract.py to specify a minimum threshold percentage of overlap (by default 5%) products must maintain wrt the specified bounding box. If no bounding box is specified, all products must maintain this threshold instead wrt the longest scene in the stack. References to products that fail to meet this criteria are removed.

Rob, I didn't have a chance to test this out with GACOS corrections. Could you give that a try, in addition to just playing around with the proposed options?

Commit:
Support for downsampling and kicking out short scenes

sssangha committed:
Bug fix in downsampling

If downsampling were specified, only one of the unw or con_comp files were downsampled, not both. In the "export_product" function, each layer is looped through and once either of these layers (e.g. unw phase) is referenced then both unw/con_comp files are mosaicked/cropped and a conditional prevents duplicating this effort, such that in the next loop when the other layer is referenced (e.g. con_comp) no work is done because again both of these layers have already been extracted.
Thus, I had to also make sure that both layers were resampled together.

Sort of convoluted, I know, so don't hesitate to take a look at the changes and logic of the code, or ask me for clarification.

sssangha committed:
remove trailing whitespace

dbekaert commented:
@sssangha not sure if you might have captured earlier discussions with @rzinke.

Can you clarify:
- Does the percentage downsampling imply an integer multiple of the resolution?
- The percentage overlap might not be a great constraint as it does not describe the overlap in square km area. i.e. This is not the same threshold for a small compared to a larger AOI. Perhaps we can convert it to a metric unit threshold. 
- I am unclear why the evaluation needs to be done per product and not per merged product.
e.g. keep all products which covers the bbox, and ensure that the coverage of the merged product meets the requirement.

dbekaert commented on tools/ARIAtools/extractProduct.py:
```diff
@@ -294,22 +300,33 @@ def merged_productbbox(product_dict, workdir='./', bbox_file=None, croptounion=F
+                rejected_scenes.append(product_dict.index(scene))
+                os.remove(scene['productBoundingBox'][0])
+            else:
+                per_overlap=(prods_bbox.area/total_bbox.area)*100
```
Originally this is an area, but gets converted to a percentage. An area threshold would be something that can be related across AOI's provided. Threshold based on percentage becomes a bit more AOI specific.

sssangha replied:
But this percentage is then directly compared to the user-defined percentage of overlap. Are you suggesting then instead to have the user-defined input in units of m^2 or km^2?

dbekaert replied:
Im referring to a threshold that can be applied across different AOI's and have a physical meaning. The percentage requires knowledge in advance on the AOI extend. E.g. the maximum union etc could change if there is a poor product being generated.

rzinke replied:
I agree, if this is practical. I had originally kicked out products based on fraction of the frame area in part for this reason... because the frame areas are relatively consistent, but one doesn't need to know the exact area of a frame in km2. 

But otherwise, I follow your logic that it is better to implement this criterion in the lower part of the script, after the merged area has been defined.

dbekaert replied:
Frames are only consistent within AOI, but not on AOI boudaries.

sssangha replied:
Yeah, kind of straying away from your point for a sec, it seems you would need to have a rough idea of the AOI extent a priori. I've already started to have problems for instance with the CA study area because track 87 exists in both CA and Saudi Arabia.

So I'm not sure of the best way to proceed without at least specifying a rough bbox.

Currently if no bbox is specified, the code takes the longest scene and uses it as a spatial reference (i.e. its area becomes the "bbox"), see lines 289-292.

dbekaert commented on tools/ARIAtools/extractProduct.py:
```diff
@@ -294,22 +300,33 @@ def merged_productbbox(product_dict, workdir='./', bbox_file=None, croptounion=F
-        # Estimate percentage of overlap with bbox
-        per_overlap=((user_bbox.area)/(open_shapefile(bbox_file, 0, 0).area))*100
-        if per_overlap<50.:
-            print("WARNING: Common track extent only has %d%% overlap with bbox"%per_overlap+'\n')
```
At the product stage once could keep it to only products that cover the bbox.
At the merge stage one could compare the threshold of coverage.
The latter is now removed.

sssangha replied:
Yeah, in this function we still compare the overlap of the merged product, not of each individual product. The logic here has been moved up to lines ~305-314

dbekaert replied:
ok thanks for confirming. The comparison on the merged box only is good. Note that that is different from the explanation given in the  PR description.

dbekaert commented on tools/ARIAtools/vrtmanager.py:
```diff
@@ -66,3 +66,31 @@ def renderOGRVRT(vrt_filename, src_datasets):
+    # Must resample mask/connected components files with nearest-neighbor
+    if fname.split('/')[-2]=='mask' or fname.split('/')[-2]=='connectedComponents':
+        # Resample raster
+        gdal.Warp(fname, fname, options=gdal.WarpOptions(format=outputFormat, cutlineDSName=prods_TOTbbox, outputBounds=bounds, width=arrshape[1], height=arrshape[0], resampleAlg='near',multithread=True, options=['NUM_THREADS=%s'%(num_threads)+' -overwrite']))
```
Ok nearest for connected component and mask

sssangha commented:
> @sssangha not sure if you might have captured earlier discussions with @rzinke.
> 
> Can you clarify:
> 
>     * Does the percentage downsampling imply an integer multiple of the resolution?
> 
>     * The percentage overlap might not be a great constraint as it does not describe the overlap in square km area. i.e. This is not the same threshold for a small compared to a larger AOI. Perhaps we can convert it to a metric unit threshold.
> 
>     * I am unclear why the evaluation needs to be done per product and not per merged product.
>       e.g. keep all products which covers the bbox, and ensure that the coverage of the merged product meets the requirement.



> @sssangha not sure if you might have captured earlier discussions with @rzinke.
> 
> Can you clarify:
> 
>     * Does the percentage downsampling imply an integer multiple of the resolution?
> 

@dbekaert Yeah, I just realized the logic of the code as is was too basic as it directly assigned the resampled x/y dimensions (e.g. original raster dimensions * percentage), which doesn't ensure the aspect ratio remains locked. So instead, I am pushing a change that specifies the resolution directly, see the changes in the vrtmanager

sssangha committed:
Update vrtmanager

Set resolution directly in order to ensure fixed aspect ratio.

dbekaert commented on tools/ARIAtools/vrtmanager.py:
```diff
@@ -79,17 +79,16 @@ def resampleRaster(fname, cellResolution, bounds, prods_TOTbbox, outputFormat='E
     ds=gdal.Warp('', fname, options=gdal.WarpOptions(format="MEM", outputBounds=bounds, multithread=True, options=['NUM_THREADS=%s'%(num_threads)]))
-    arrshape=[ds.RasterYSize, ds.RasterXSize] # Get shape of full res layers
-    arrshape=[int(i*(cellResolution/100.)) for i in arrshape]
+    arrshape=[abs(ds.GetGeoTransform()[1])/(cellResolution/100.), abs(ds.GetGeoTransform()[-1])/(cellResolution/100.)] # Get output res
```
To confirm, this does not imply integer multiple of the resolution.

sssangha replied:
This isn't necessarily an integer multiple of the natural resolution, but the aspect ratio is preserved, and consistent vis-a-vis the other standard product layers and DEM/mask. 

I'm not sure how to provide more input flexibility than allowing to specify a percentage. E.g. if we force the input to be an integer multiple, I'm not sure of what we can allow for outside of specifying 1-arc,2-arc,4-arc, etc.

sssangha committed:
vrtmanager bugfix

Fixed to not force user to generate raster if they specified VRT format. Now the VRT file itself is updated directly and no raster generated, unless specified.

sssangha commented:
@rzinke Ok, check out the latest fix I made to the vrtmanager. Now it handles cases where users just wish to generate VRT files.

rzinke commented:
Hi @sssangha thanks for all the modifications. I am still getting an error when I try to use the downsampling routine in ariaTSsetup. Here's the error: 

7 out of 30 interferograms rejected since stitched interferogram would have gaps
Thread count specified for gdal multiprocessing = 8
Download/cropping DEM
Downloaded 3 arc-sec SRTM DEM here: TSextract2/DEM/SRTM_3arcsec.dem

Extracting unwrapped phase, coherence, and connected components for each interferogram pair
Generating: unwrappedPhase - [======================= 83% =============>        ] 20181128_20181116  634s /   129sMulti-core version failed, will try single for loop
Generating: unwrappedPhase - [==================================================] 20181222_20181210  758s /    31s
Generating: coherence - [==================================================] 20181222_20181210  262s /    10s

Extracting single incidence angle, look angle and azimuth angle files valid over common interferometric grid
Generating: incidenceAngle - [==================================================] 20180707_20180613
Generating: lookAngle - [==================================================] 20180707_20180613
Generating: azimuthAngle - [==================================================] 20180707_20180613
Creating directory: TSextract2/stack
Number of coherence discovered:  23
Traceback (most recent call last):
  File "/u/sar-r0/rzinke/python/miniconda3/envs/ARIA-tools/bin/ariaTSsetup.py", line 15, in <module>
    main(inps)
  File "/u/sar-r0/rzinke/python/miniconda3/envs/ARIA-tools/lib/python3.7/site-packages/ARIAtools/tsSetup.py", line 316, in main
    generateStack(standardproduct_info,'coherence','cohStack',workdir=inps.workdir)
  File "/u/sar-r0/rzinke/python/miniconda3/envs/ARIA-tools/lib/python3.7/site-packages/ARIAtools/tsSetup.py", line 147, in generateStack
    ds = gdal.Open(data, gdal.GA_ReadOnly)
  File "/u/sar-r0/rzinke/python/miniconda3/envs/ARIA-tools/lib/python3.7/site-packages/osgeo/gdal.py", line 3308, in Open 
    return _gdal.Open(*args)
RuntimeError: GDALWarpOptions.Validate(): hSrcDS is not set.

Commit:
Multilooking factor, overlap computed for area, bug fixes

sssangha commented:
@rzinke @dbekaert listed are the updates I have pushed:

When downsampling (with the --multilooking flag), users now have to specify an integer number instead of a percentage. I.e. multilooking factor is an integer multiple of standard resolution. For example 2 = 90m*2 = 180m. 

Threshold of coverage now controlled by km^2 of overlap, as opposed to percentage of overlap. Equal area projection used.

Bug-fix: If mask specified/provided as input, the connected component file masked values are now assigned as -1 as opposed to 0, in agreement with the connected component nodata values.

Bug-fix: Error resolved when user specifies VRT output format and wishes to downsample outputs.

sssangha committed:
Remove print statement used for debugging

Commit:
VRT bugfix

sssangha committed:
codacy-suggested fixes

dbekaert commented:
@sssangha @rzinke i noticed this PR is still open. Did we get closure on it?

sssangha commented:
@dbekaert @rzinke 
Refer to my new "rankedresampleVSdefaultresample" branch where I merge the functionality of the two existing branches.

I will leave open the other two aforementioned branches for reference and will delete them once these changes are pushed.

rzinke commented:
Hi @sssangha, thanks! Could you remind me the difference between ranked resample and default resample? 

I can start testing these this afternoon.

sssangha commented:
Hello @rzinke, the difference between ranked resample is to determine the mode of the connected component within a given resampling window and only average phase pixels spatially coincident with said mode. In cases where there are multiple modes (e.g. an equal number of the two most frequent connected component pixels), the smallest value is selected.

The default resampling doesn't take into account the connected components and just does a simple average in a given window.
